scripts/jsonrpcproxy.py

```diff
@@ -31,13 +32,71 @@
 DELIMITER = ord('\n')
 INFO = """JSON-RPC Proxy
 
-Version: {}
-Proxy: {}
-Backend: {}
+Version:  {version}
+Proxy:    {proxy_url}
+Backend:  {backend_url} (connected: {connected})
 """
 
 
-class NamedPipe(object):
+class BackendError(Exception):
+    pass
+
+
+class UnixSocketConnector(object):
+    """Unix Domain Socket connector. Connects to socket lazily."""
+
+    def __init__(self, socket_path):
+        self.socket_path = socket_path
+        self.socket = None
+
+    @staticmethod
+    def _get_error_message(os_error_number):
+        if os_error_number == errno.ENOENT:
+            return "Unix Domain Socket '{}' does not exist"
+        if os_error_number == errno.ECONNREFUSED:
+            return "Connection to '{}' refused"
+        return "Unknown error when connecting to '{}'"
+
+    def is_connected(self):
+        return self.socket is not None
+
+    def _connect(self):
+        if self.socket is None:
+            try:
+                s = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
+                s.connect(self.socket_path)
+                s.settimeout(1)
+                # Assign last, to keep it None in case of exception.
+                self.socket = s
+            except OSError as ex:
+                msg = self._get_error_message(ex.errno)
+                err = BackendError(msg.format(self.socket_path))
+                raise err from ex
+
+    def _reconnect(self):
+        self.socket.shutdown(socket.SHUT_RDWR)
+        self.socket.close()
+        self.socket = None
+        self._connect()
+
+    def recv(self, max_length):
+        self._connect()
```

Comment:
is it possible in python instead of explicitly calling this to have `get_socket()` method that will return `self.socket` intitializing it if needed ?

Reply:
You mean `self.get_socket().recv(...)`? Yes. You can even have a property like this `self.socket.recv(...)`, but I'd like to avoid Pytonish features.

I will try this, let's see how it looks.